power-sequencer/mihawk-cpld.cpp

```diff
@@ -0,0 +1,478 @@
+        }
+        else if (errorcode == 29)
+        {
+            report<PowerOnErrorVcs1PGOODFail>();
+            std::cerr << "Power_on error : Vcs1_pgood fail! \n";
+        }
+        else if (errorcode == 30)
+        {
+            report<PowerOnErrorVddr0PGOODFail>();
+            std::cerr << "Power_on error : Vddr0_pgood fail! \n";
+        }
+        else if (errorcode == 31)
+        {
+            report<PowerOnErrorVtt0PGOODFail>();
+            std::cerr << "Power_on error : Vtt0_pgood fail! \n";
+        }
+        else if (errorcode == 32)
+        {
+            report<PowerOnErrorVddr1PGOODFail>();
+            std::cerr << "Power_on error : Vddr1_pgood fail! \n";
+        }
+        else if (errorcode == 33)
+        {
+            report<PowerOnErrorVtt1PGOODFail>();
+            std::cerr << "Power_on error : Vtt1_pgood fail! \n";
+        }
+        else if (errorcode == 34)
+        {
+            report<PowerOnErrorGPU0PGOODFail>();
+            std::cerr << "Power_on error : GPU0_pgood fail! \n";
+        }
+        else if (errorcode == 35)
+        {
+            report<PowerOnErrorGPU1PGOODFail>();
+            std::cerr << "Power_on error : GPU1_pgood fail! \n";
+        }
+        else if (errorcode == 170)
+        {
+            report<PowerOnErrorPSU0PSU1PGOODFail>();
+            std::cerr << "Power_on error : Psu0&PSU1_pgood fail! \n";
+        }
+    }
+}
+
+void MihawkCPLD::analyze()
+{
+    bool psuError = checkPSUdevice();
+    if (psuError)
+    {
+        // If PSU not present, report the error log & event.
+        report<PSUPresentError>();
+        std::cerr << "CPLD : PSU Present Error! \n";
+    }
+}
+
+// Read CPLD_register error code and return the result to analyze.
+int MihawkCPLD::readFromCPLDPSUErrorCode(int bus, int Addr)
+{
+    std::string i2cBus = "/dev/i2c-" + std::to_string(bus);
+
+    // open i2c device(CPLD-PSU-register table)
+    int fd = open(i2cBus.c_str(), O_RDWR | O_CLOEXEC);
+    if (fd < 0)
+    {
+        std::cerr << "Unable to open i2c device \n";
+    }
+
+    // set i2c slave address
+    if (ioctl(fd, I2C_SLAVE_FORCE, Addr) < 0)
+    {
+        std::cerr << "Unable to set device address \n";
+        close(fd);
+    }
+
+    // check whether support i2c function
+    unsigned long funcs = 0;
+    if (ioctl(fd, I2C_FUNCS, &funcs) < 0)
+    {
+        std::cerr << "Not support I2C_FUNCS \n";
+        close(fd);
+    }
+
+    // check whether support i2c-read function
+    if (!(funcs & I2C_FUNC_SMBUS_READ_BYTE_DATA))
+    {
+        std::cerr << "Not support I2C_FUNC_SMBUS_READ_BYTE_DATA \n";
+        close(fd);
+    }
+
+    int statusValue_2;
+    unsigned int statusReg_2 = StatusReg_2;
+
+    statusValue_2 = i2c_smbus_read_byte_data(fd, statusReg_2);
+    close(fd);
+
+    if (statusValue_2 < 0)
+    {
+        statusValue_2 = 0;
+    }
+
+    // return the i2c-read data
+    return statusValue_2;
+}
+
+// Check PSU_present_pin.
+bool MihawkCPLD::checkPSUdevice()
```

Comment:
I will add this part of the reason because there is also one field in Mihawk's CPLD-register table that can be read and determine if PSU0 and PSU1 are "present".

I know that this part of the function already has a similar function in the dbus-monitor to determine if the PSU is "present".

However, my idea is that this API operation is based on pgood exceptions (not changed from 0 to 1) when chassis power on, as long as the correlation will cause the power to boot abnormally, the cause should be recorded.

Reply:
@spinler  Ok, I understand what you mean, I will re-modify this part.